File: ros2_ws/src/ai_inference/ai_inference/heatmap_bias_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
import numpy as np
from std_msgs.msg import Float32MultiArray

# === Only these 2 lines changed ===
GRID_SIZE = 128          # 256 → 128
T_OUT = 16               # 10 → 16
# ==================================

NUM_SECTORS = 72
ALPHA = 3.0
CELL_SIZE = 0.1
DIST_DECAY = 0.4


class HeatmapBiasNode(Node):
    def __init__(self):
        super().__init__("heatmap_bias_node")

        self.H = GRID_SIZE
        self.W = GRID_SIZE
        self.cy = self.H // 2
        self.cx = self.W // 2

        self.max_ray_steps = min(self.H, self.W) // 2

        self.sub = self.create_subscription(
            Float32MultiArray,
            "/future_occupancy",
            self.on_future_occupancy,
            10
        )
        self.pub_risk = self.create_publisher(
            Float32MultiArray,
            "/future_bias/risk_map",
            10
        )
        self.pub_omega = self.create_publisher(
            Float32MultiArray,
            "/future_bias/omega_weights",
            10
        )

        self.get_logger().info("🔥 Heatmap Bias Node (128×128, T_OUT=16) Started")


    def on_future_occupancy(self, msg: Float32MultiArray):
        data = np.array(msg.data, dtype=np.float32)
        expected = T_OUT * self.H * self.W

        if data.size != expected:
            self.get_logger().warn(
                f"⚠ future_occupancy size mismatch: got {data.size}, expected {expected}"
            )
            return

        future = data.reshape(T_OUT, self.H, self.W)

        # === Only this line changed ===
        mid_indices = [4,5,6,7,8,9,10]    # 기존 로직 유지 + 길이에 맞게 조정
        # ===============================

        mid_frames = future[mid_indices]

        weights = np.array([0.3, 0.25, 0.2, 0.15, 0.1, 0.05, 0.05], dtype=np.float32)

        risk_map = np.tensordot(weights, mid_frames, axes=([0], [0]))

        msg_r = Float32MultiArray()
        msg_r.data = risk_map.flatten().tolist()
        self.pub_risk.publish(msg_r)

        # === compute_sector_risk + risk_to_weights: NO CHANGES ===
        sector_risk = self.compute_sector_risk(risk_map)
        omega = self.risk_to_weights(sector_risk)

        msg_w = Float32MultiArray()
        msg_w.data = omega.astype(np.float32).tolist()
        self.pub_omega.publish(msg_w)


    # === NO CHANGE BELOW ===
    def compute_sector_risk(self, risk_map):
        sector_vals = np.zeros(NUM_SECTORS, dtype=np.float32)
        two_pi = 2.0 * np.pi

        for i in range(NUM_SECTORS):
            angle_center = -np.pi + (i+0.5) * (two_pi / NUM_SECTORS)
            cos_a = np.cos(angle_center)
            sin_a = np.sin(angle_center)

            ray_max = 0.0

            for step in range(2, self.max_ray_steps):
                row = self.cy - step * cos_a
                col = self.cx - step * sin_a

                iy = int(round(row))
                ix = int(round(col))

                if ix < 0 or ix >= self.W or iy < 0 or iy >= self.H:
                    break

                base = risk_map[iy, ix]
                if base <= 0:
                    continue

                dist_m = step * CELL_SIZE
                atten = np.exp(-DIST_DECAY * dist_m)
                eff = float(base * atten)

                if eff > ray_max:
                    ray_max = eff

            sector_vals[i] = ray_max

        return sector_vals


    def risk_to_weights(self, sector_risk):
        weights = np.exp(-ALPHA * sector_risk)
        max_w = np.max(weights)
        if max_w < 1e-6:
            weights[:] = 1.0
        else:
            weights /= max_w
        return weights


def main(args=None):
    rclpy.init(args=args)
    node = HeatmapBiasNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros2_ws/src/ai_inference_cpp/src/heatmap_bias_node.cpp
#include <memory>
#include <vector>
#include <cmath>
#include <algorithm>
#include <cstring> // for memcpy

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float32_multi_array.hpp"
#include "std_msgs/msg/u_int8_multi_array.hpp" // [중요] UInt8 사용

using std::placeholders::_1;
using std::vector;

static constexpr int GRID_SIZE   = 128;
static constexpr int T_OUT       = 16;
static constexpr int NUM_SECTORS = 72;

static constexpr float ALPHA      = 3.0f;
static constexpr float CELL_SIZE  = 0.1f;
static constexpr float DIST_DECAY = 0.4f;

// 가중치 인덱스
static constexpr int MID_START = 4;
static constexpr int MID_END   = 10;
static constexpr int MID_COUNT = (MID_END - MID_START + 1);

class HeatmapBiasNode : public rclcpp::Node
{
public:
  HeatmapBiasNode()
  : Node("heatmap_bias_node"),
    H_(GRID_SIZE),
    W_(GRID_SIZE),
    cy_(GRID_SIZE / 2),
    cx_(GRID_SIZE / 2),
    max_ray_steps_(std::min(GRID_SIZE, GRID_SIZE) / 2)
  {
    // [수정] UInt8MultiArray 구독
    sub_ = this->create_subscription<std_msgs::msg::UInt8MultiArray>(
      "/future_occupancy", 10,
      std::bind(&HeatmapBiasNode::onFutureOccupancy, this, _1));

    pub_risk_ = this->create_publisher<std_msgs::msg::Float32MultiArray>(
      "/future_bias/risk_map", 10);

    pub_omega_ = this->create_publisher<std_msgs::msg::Float32MultiArray>(
      "/future_bias/omega_weights", 10);

    RCLCPP_INFO(this->get_logger(),
      "🔥 Heatmap Bias Node (Optimized UInt8 Input) Started");
  }

private:
  void onFutureOccupancy(const std_msgs::msg::UInt8MultiArray::SharedPtr msg)
  {
    // 1. 데이터 검증 (Byte Size 체크)
    const size_t byte_size = msg->data.size();
    const size_t expected_floats = static_cast<size_t>(T_OUT) * H_ * W_;
    const size_t expected_bytes = expected_floats * sizeof(float);

    if (byte_size != expected_bytes) {
      RCLCPP_WARN(this->get_logger(),
        "Size mismatch! Got %zu bytes, expected %zu bytes",
        byte_size, expected_bytes);
      return;
    }

    // 2. [핵심] 바이트 포인터를 Float 포인터로 재해석 (Zero-copy logic)
    const float* raw_data = reinterpret_cast<const float*>(msg->data.data());

    // 3. Risk Map 계산 (이전과 로직 동일)
    vector<float> risk_map(H_ * W_, 0.0f);
    const float weights[MID_COUNT] = {0.3f, 0.25f, 0.2f, 0.15f, 0.1f, 0.05f, 0.05f};

    // raw_data[idx]는 float처럼 접근 가능
    for (int y = 0; y < H_; ++y) {
      for (int x = 0; x < W_; ++x) {
        float acc = 0.0f;
        int w_idx = 0;
        for (int t = MID_START; t <= MID_END; ++t, ++w_idx) {
          size_t idx = static_cast<size_t>(t) * H_ * W_
                     + static_cast<size_t>(y) * W_
                     + static_cast<size_t>(x);
          acc += weights[w_idx] * raw_data[idx];
        }
        risk_map[y * W_ + x] = acc;
      }
    }

    // 4. 섹터 위험도 계산
    vector<float> sector_risk(NUM_SECTORS, 0.0f);
    computeSectorRisk(risk_map, sector_risk);

    // 5. 가중치 변환
    vector<float> omega(NUM_SECTORS, 1.0f);
    riskToWeights(sector_risk, omega);

    // 6. Publish (결과는 다시 Float32로 보냄 - 시각화용)
    std_msgs::msg::Float32MultiArray risk_msg;
    risk_msg.data = risk_map; 
    pub_risk_->publish(risk_msg);

    std_msgs::msg::Float32MultiArray omega_msg;
    omega_msg.data = omega;
    pub_omega_->publish(omega_msg);
  }

  void computeSectorRisk(const vector<float> & risk_map,
                         vector<float> & sector_vals)
  {
    const float two_pi = 2.0f * static_cast<float>(M_PI);

    for (int i = 0; i < NUM_SECTORS; ++i) {
      float angle_center = -static_cast<float>(M_PI)
                         + (static_cast<float>(i) + 0.5f) * (two_pi / NUM_SECTORS);
      float cos_a = std::cos(angle_center);
      float sin_a = std::sin(angle_center);

      float ray_max = 0.0f;

      for (int step = 2; step < max_ray_steps_; ++step) {
        float row = static_cast<float>(cy_) - step * cos_a;
        float col = static_cast<float>(cx_) - step * sin_a;

        int iy = static_cast<int>(std::round(row));
        int ix = static_cast<int>(std::round(col));

        if (ix < 0 || ix >= W_ || iy < 0 || iy >= H_) break;

        float base = risk_map[iy * W_ + ix];
        if (base <= 0.0f) continue;

        float dist_m = static_cast<float>(step) * CELL_SIZE;
        float eff    = base * std::exp(-DIST_DECAY * dist_m);

        if (eff > ray_max) ray_max = eff;
      }
      sector_vals[i] = ray_max;
    }
  }

  void riskToWeights(const vector<float> & sector_risk,
                     vector<float> & weights)
  {
    float max_w = 0.0f;
    for (int i = 0; i < NUM_SECTORS; ++i) {
      float w = std::exp(-ALPHA * sector_risk[i]);
      weights[i] = w;
      if (w > max_w) max_w = w;
    }

    if (max_w < 1e-6f) {
      std::fill(weights.begin(), weights.end(), 1.0f);
      return;
    }
    for (float & w : weights) {
      w /= max_w;
    }
  }

private:
  int H_, W_, cy_, cx_, max_ray_steps_;
  
  // [수정] 구독 타입
  rclcpp::Subscription<std_msgs::msg::UInt8MultiArray>::SharedPtr sub_;
  
  rclcpp::Publisher<std_msgs::msg::Float32MultiArray>::SharedPtr pub_risk_;
  rclcpp::Publisher<std_msgs::msg::Float32MultiArray>::SharedPtr pub_omega_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<HeatmapBiasNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}